File: src/lib/crypto/x3dh.ts
// XARK OS v2.0 — X3DH Key Agreement
// Extended Triple Diffie-Hellman for establishing 1:1 sessions.
// Used to initialize Double Ratchet sessions.

import { dh, hkdf, verify, ed25519PkToCurve25519, generateDHKeyPair } from './primitives';
import type { PublicKeyBundle, RawKeyPair } from './types';

const X3DH_INFO = 'XarkE2EE-x3dh';

/**
 * Initiator side of X3DH — creates shared secret for session setup.
 * Called by the person sending the first message.
 */
export function x3dhInitiate(
  myIdentityCurve25519: { publicKey: Uint8Array; privateKey: Uint8Array },
  peerBundle: PublicKeyBundle
): { sharedSecret: Uint8Array; ephemeralKey: RawKeyPair } {
  // BUG 11/13 hardening: validate initiator identity key
  if (!myIdentityCurve25519.privateKey.length || !myIdentityCurve25519.publicKey.length) {
    throw new Error('X3DH: Invalid initiator identity key — zero-length key material');
  }

  if (!peerBundle.identityKey.length || !peerBundle.signedPreKey.length) {
    throw new Error('X3DH: Invalid zero-length keys provided');
  }

  // Verify signed pre-key signature
  const isValid = verify(
    peerBundle.preKeySig,
    peerBundle.signedPreKey,
    peerBundle.identityKey // Ed25519 public key
  );
  if (!isValid) throw new Error('X3DH: Invalid signed pre-key signature');

  // Convert peer's Ed25519 identity key to Curve25519
  const peerIdentityCurve = ed25519PkToCurve25519(peerBundle.identityKey);

  // Generate ephemeral key pair
  const ephemeralKey = generateDHKeyPair();

  // Four DH computations
  const dh1 = dh(myIdentityCurve25519.privateKey, peerBundle.signedPreKey);   // IK_A x SPK_B
  const dh2 = dh(ephemeralKey.privateKey, peerIdentityCurve);                  // EK_A x IK_B
  const dh3 = dh(ephemeralKey.privateKey, peerBundle.signedPreKey);            // EK_A x SPK_B

  let ikm: Uint8Array;
  if (peerBundle.oneTimePreKey) {
    const dh4 = dh(ephemeralKey.privateKey, peerBundle.oneTimePreKey);         // EK_A x OTK_B
    ikm = concatBytes(dh1, dh2, dh3, dh4);
  } else {
    ikm = concatBytes(dh1, dh2, dh3);
  }

  const sharedSecret = hkdf(ikm, new Uint8Array(32), X3DH_INFO, 32);
  return { sharedSecret, ephemeralKey };
}

/**
 * Responder side of X3DH — derives same shared secret from initiator's message.
 * Called when receiving the first message in a session.
 */
export function x3dhRespond(
  myIdentityCurve25519: { publicKey: Uint8Array; privateKey: Uint8Array },
  mySignedPreKey: RawKeyPair,
  myOneTimePreKey: RawKeyPair | null,
  peerIdentityCurve25519Public: Uint8Array,
  peerEphemeralPublic: Uint8Array
): Uint8Array {
  // BUG 11 hardening: reject missing ephemeral key — cannot compute shared secret without it
  if (!peerEphemeralPublic || !peerEphemeralPublic.length) {
    throw new Error('X3DH: Missing peer ephemeral key — cannot compute shared secret');
  }
  if (!peerIdentityCurve25519Public || !peerIdentityCurve25519Public.length) {
    throw new Error('X3DH: Missing peer identity key');
  }
  if (!mySignedPreKey.privateKey.length) {
    throw new Error('X3DH: Invalid signed pre-key — zero-length private key');
  }

  const dh1 = dh(mySignedPreKey.privateKey, peerIdentityCurve25519Public);    // SPK_B x IK_A
  const dh2 = dh(myIdentityCurve25519.privateKey, peerEphemeralPublic);       // IK_B x EK_A
  const dh3 = dh(mySignedPreKey.privateKey, peerEphemeralPublic);             // SPK_B x EK_A

  let ikm: Uint8Array;
  if (myOneTimePreKey) {
    const dh4 = dh(myOneTimePreKey.privateKey, peerEphemeralPublic);          // OTK_B x EK_A
    ikm = concatBytes(dh1, dh2, dh3, dh4);
  } else {
    ikm = concatBytes(dh1, dh2, dh3);
  }

  return hkdf(ikm, new Uint8Array(32), X3DH_INFO, 32);
}

/** Concatenate multiple Uint8Arrays */
function concatBytes(...arrays: Uint8Array[]): Uint8Array {
  const total = arrays.reduce((sum, a) => sum + a.length, 0);
  const result = new Uint8Array(total);
  let offset = 0;
  for (const arr of arrays) {
    result.set(arr, offset);
    offset += arr.length;
  }
  return result;
}

File: engine/lib/src/crypto/x3dh/x3dh.dart
import 'dart:typed_data';

import 'package:hello_engine/src/crypto/keys/ed25519_to_curve25519.dart';
import 'package:hello_engine/src/crypto/keys/key_types.dart';
import 'package:cryptography/cryptography.dart';

/// Base exception for X3DH protocol errors.
class X3dhException implements Exception {
  /// Creates an [X3dhException] with the given [message].
  X3dhException(this.message);

  /// Human-readable description of the error.
  final String message;

  @override
  String toString() => 'X3dhException: $message';
}

/// Thrown when no one-time pre-keys are available.
///
/// This is informational — the protocol falls back to 3-DH
/// (which is valid Signal behavior).
class OtkExhaustedException extends X3dhException {
  /// Creates an [OtkExhaustedException].
  OtkExhaustedException()
      : super('No one-time pre-keys available — falling back to 3-DH');
}

/// Thrown when the signed pre-key signature fails verification.
///
/// This may indicate a man-in-the-middle attack.
class SignatureVerificationFailed extends X3dhException {
  /// Creates a [SignatureVerificationFailed].
  SignatureVerificationFailed()
      : super(
          'Signed pre-key signature verification failed — potential MITM',
        );
}

/// Thrown when identity key in the bundle does not match expectations.
///
/// Typically means the cached bundle is stale or has been tampered with.
class KeyMismatchError extends X3dhException {
  /// Creates a [KeyMismatchError].
  KeyMismatchError()
      : super('Key bundle mismatch — cached bundle may be stale');
}

/// Warning exception for stale signed pre-keys (older than 7 days).
///
/// The protocol proceeds, but callers should log this condition.
class StalePreKeyWarning extends X3dhException {
  /// Creates a [StalePreKeyWarning].
  StalePreKeyWarning()
      : super(
          'Signed pre-key is older than 7 days — proceeding with warning',
        );
}

/// Extended Triple Diffie-Hellman (X3DH) key agreement.
///
/// Implements the X3DH protocol for establishing shared secrets
/// between two parties. Uses Ed25519 identity keys converted to
/// X25519 for DH, with the HKDF info string `'XarkE2EE-x3dh'`
/// and 32-byte zero salt.
class X3DH {
  X3DH._();

  /// HKDF info string for X3DH key derivation.
  static const String _hkdfInfo = 'XarkE2EE-x3dh';

  /// Performs the X3DH key agreement as the initiator (Alice).
  ///
  /// 1. Verifies the signed pre-key signature.
  /// 2. Converts Ed25519 identity keys to X25519.
  /// 3. Generates an ephemeral X25519 key pair.
  /// 4. Computes DH1..DH4 and derives a 32-byte shared secret.
  static Future<X3DHResult> initiatorKeyAgreement({
    required IdentityKeyPair ourIdentity,
    required PreKeyBundle theirBundle,
  }) async {
    // 1. Verify the signed pre-key signature.
    final ed = Ed25519();
    final sigValid = await ed.verify(
      theirBundle.signedPreKey,
      signature: Signature(
        theirBundle.preKeySignature,
        publicKey: SimplePublicKey(
          theirBundle.identityKey,
          type: KeyPairType.ed25519,
        ),
      ),
    );
    if (!sigValid) {
      throw SignatureVerificationFailed();
    }

    // 2. Convert their Ed25519 identity key to X25519.
    final theirX25519Pub = Ed25519ToCurve25519.convertPublicKey(
      theirBundle.identityKey,
    );

    // 3. Generate ephemeral X25519 key pair.
    final x = X25519();
    final ephKp = await x.newKeyPair();
    final ephEx = await ephKp.extract();
    final ephPub = Uint8List.fromList(ephEx.publicKey.bytes);
    final ephPriv = Uint8List.fromList(ephEx.bytes);

    // 4. Compute the four DH operations.
    final dh1 = await _dh(
      x,
      ourIdentity.x25519PrivateKey,
      ourIdentity.x25519PublicKey,
      theirBundle.signedPreKey,
    );
    final dh2 = await _dh(
      x,
      ephPriv,
      ephPub,
      theirX25519Pub,
    );
    final dh3 = await _dh(
      x,
      ephPriv,
      ephPub,
      theirBundle.signedPreKey,
    );

    Uint8List? dh4;
    if (theirBundle.oneTimePreKey != null) {
      dh4 = await _dh(
        x,
        ephPriv,
        ephPub,
        theirBundle.oneTimePreKey!,
      );
    }

    // 5. Concatenate all DH outputs.
    final ikmLen = 32 * 3 + (dh4 != null ? 32 : 0);
    final ikm = Uint8List(ikmLen)
      ..setAll(0, dh1)
      ..setAll(32, dh2)
      ..setAll(64, dh3);
    if (dh4 != null) {
      ikm.setAll(96, dh4);
    }

    // 6. HKDF to derive the 32-byte shared secret.
    final sharedSecret = await _deriveSecret(ikm);

    return X3DHResult(
      sharedSecret: sharedSecret,
      ephemeralPublicKey: ephPub,
      usedOneTimePreKeyId: theirBundle.oneTimePreKeyId,
    );
  }

  /// Performs the X3DH key agreement as the responder (Bob).
  ///
  /// Mirror of [initiatorKeyAgreement] with swapped roles.
  static Future<X3DHResult> responderKeyAgreement({
    required IdentityKeyPair ourIdentity,
    required SignedPreKey ourSignedPreKey,
    required OneTimePreKey? ourOneTimePreKey,
    required Uint8List theirIdentityKey,
    required Uint8List theirEphemeralKey,
  }) async {
    final theirX25519Pub = Ed25519ToCurve25519.convertPublicKey(
      theirIdentityKey,
    );
    final x = X25519();

    final dh1 = await _dh(
      x,
      ourSignedPreKey.privateKey,
      ourSignedPreKey.publicKey,
      theirX25519Pub,
    );
    final dh2 = await _dh(
      x,
      ourIdentity.x25519PrivateKey,
      ourIdentity.x25519PublicKey,
      theirEphemeralKey,
    );
    final dh3 = await _dh(
      x,
      ourSignedPreKey.privateKey,
      ourSignedPreKey.publicKey,
      theirEphemeralKey,
    );

    Uint8List? dh4;
    if (ourOneTimePreKey != null) {
      dh4 = await _dh(
        x,
        ourOneTimePreKey.privateKey,
        ourOneTimePreKey.publicKey,
        theirEphemeralKey,
      );
    }

    final ikmLen = 32 * 3 + (dh4 != null ? 32 : 0);
    final ikm = Uint8List(ikmLen)
      ..setAll(0, dh1)
      ..setAll(32, dh2)
      ..setAll(64, dh3);
    if (dh4 != null) {
      ikm.setAll(96, dh4);
    }

    final sharedSecret = await _deriveSecret(ikm);

    return X3DHResult(
      sharedSecret: sharedSecret,
      ephemeralPublicKey: theirEphemeralKey,
      usedOneTimePreKeyId: ourOneTimePreKey?.id,
    );
  }

  /// Performs an X25519 Diffie-Hellman operation.
  static Future<Uint8List> _dh(
    X25519 algorithm,
    Uint8List ourPrivate,
    Uint8List ourPublic,
    Uint8List theirPublic,
  ) async {
    final keyPair = SimpleKeyPairData(
      ourPrivate,
      publicKey: SimplePublicKey(
        ourPublic,
        type: KeyPairType.x25519,
      ),
      type: KeyPairType.x25519,
    );
    final remotePublic = SimplePublicKey(
      theirPublic,
      type: KeyPairType.x25519,
    );

    final sharedSecret = await algorithm.sharedSecretKey(
      keyPair: keyPair,
      remotePublicKey: remotePublic,
    );
    final bytes = await sharedSecret.extractBytes();
    return Uint8List.fromList(bytes);
  }

  /// Derives a 32-byte secret from IKM via HKDF-SHA256.
  static Future<Uint8List> _deriveSecret(Uint8List ikm) async {
    final hkdf = Hkdf(
      hmac: Hmac.sha256(),
      outputLength: 32,
    );
    final derived = await hkdf.deriveKey(
      secretKey: SecretKeyData(ikm),
      nonce: Uint8List(32),
      info: _hkdfInfo.codeUnits,
    );
    return Uint8List.fromList(derived.bytes);
  }
}
